File: rerun_cpp/src/rerun/components/text_log_level_ext.cpp
#include "text_log_level.hpp"

// <CODEGEN_COPY_TO_HEADER>
#include "../rerun_sdk_export.hpp"

// </CODEGEN_COPY_TO_HEADER>

// Uncomment for better auto-complete while editing the extension.
// #define EDIT_EXTENSION

namespace rerun {
    namespace components {

#ifdef EDIT_EXTENSION
        struct TextLogLevelExt {
            std::string value;
#define TextLogLevel TextLogLevelExt

            // Don't provide a string_view constructor, std::string constructor exists and covers this.

            // <CODEGEN_COPY_TO_HEADER>

            /// Designates catastrophic failures.
            RERUN_SDK_EXPORT static const TextLogLevel Critical;

            /// Designates very serious errors.
            RERUN_SDK_EXPORT static const TextLogLevel Error;

            /// Designates hazardous situations.
            RERUN_SDK_EXPORT static const TextLogLevel Warning;

            /// Designates useful information.
            RERUN_SDK_EXPORT static const TextLogLevel Info;

            /// Designates lower priority information.
            RERUN_SDK_EXPORT static const TextLogLevel Debug;

            /// Designates very low priority, often extremely verbose, information.
            RERUN_SDK_EXPORT static const TextLogLevel Trace;

            /// Construct `TextLogLevel` from a null-terminated UTF8 string.
            TextLogLevel(const char* str) : value(str) {}

            const char* c_str() const {
                return value.c_str();
            }

            // </CODEGEN_COPY_TO_HEADER>
        };

#undef TextLogLevel
#else
#define TextLogLevelExt TextLogLevel
#endif

        const TextLogLevel TextLogLevel::Critical = "CRITICAL";
        const TextLogLevel TextLogLevel::Error = "ERROR";
        const TextLogLevel TextLogLevel::Warning = "WARN";
        const TextLogLevel TextLogLevel::Info = "INFO";
        const TextLogLevel TextLogLevel::Debug = "DEBUG";
        const TextLogLevel TextLogLevel::Trace = "TRACE";
    } // namespace components
} // namespace rerun

File: crates/store/re_sdk_types/src/components/text_log_level_ext.rs
use super::TextLogLevel;

impl TextLogLevel {
    /// Designates catastrophic failures.
    pub const CRITICAL: &'static str = "CRITICAL";

    /// Designates very serious errors.
    pub const ERROR: &'static str = "ERROR";

    /// Designates hazardous situations.
    pub const WARN: &'static str = "WARN";

    /// Designates useful information.
    pub const INFO: &'static str = "INFO";

    /// Designates lower priority information.
    pub const DEBUG: &'static str = "DEBUG";

    /// Designates very low priority, often extremely verbose, information.
    pub const TRACE: &'static str = "TRACE";

    /// The log level as a string slice, e.g. "INFO".
    #[inline]
    pub fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl From<TextLogLevel> for String {
    #[inline]
    fn from(value: TextLogLevel) -> Self {
        value.as_str().to_owned()
    }
}

impl AsRef<str> for TextLogLevel {
    #[inline]
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl std::borrow::Borrow<str> for TextLogLevel {
    #[inline]
    fn borrow(&self) -> &str {
        self.as_str()
    }
}

impl Default for TextLogLevel {
    #[inline]
    fn default() -> Self {
        Self::INFO.to_owned().into()
    }
}
